### PR TITLE
RubyGems is now an OAuth provider for *.rubygems.org sites

### PR DESCRIPTION
**Pull request by qrush**
This is the first PR as I'm sure of many for the RubyGems Adoption Center, and hopefully more apps to blossom under `*.rubygems.org` as proposed by the community. Continuing this from #1010.

A few goals here:

1. RubyGems shouldn't try to get into the "identity" game (e.g., Facebook, Google, GitHub, Twitter - etc) - so this OAuth provider is just so we can share your API credentials with approved projects
2. We need lots of docs on how to get this working and how to apply 
3. Keep it simple. OAuth is not simple. Might be an uphill battle. :grin:

### CODE REVIEW

File: config/initializers/admins.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+module Gemcutter
+  mattr_accessor :admins
+  self.admins = Application.config_for("admins")["emails"]
+end
```
Comment (bf4) [inline config/initializers/admins.rb]: Nice.  Do we want a `fetch` here for fail fast?

File: config/routes.rb [hunk 2]
```diff
@@ -149,4 +149,6 @@
   namespace :internal do
     get 'background_job_stats', to: Internal::BackgroundJobStatsController.action(:stats)
   end
+
+  use_doorkeeper scope: 'oauth'
```
Comment (bf4) [inline config/routes.rb]: Why at the bottom?

File: test/factories.rb [hunk 3]
```diff
@@ -116,4 +116,16 @@
       rubygem nil
     end
   end
+
+  factory :oauth_application, class: Doorkeeper::Application do
+    name "Adoption Center"
+    redirect_uri "https://example.org/auth"
```
Comment (bf4) [inline test/factories.rb]: Better to have it on localhost, no?

File: test/integration/authentication_test.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  test "use an Oauth access token" do
+    access = create(:oauth_access_token, resource_owner_id: @user.id)
+    authenticate "Bearer #{access.token}"
+    assert_response :success
+  end
+
+  test "use an invalid Oauth access token" do
+    authenticate "Bearer deadbeef"
```
Comment (bf4) [inline test/integration/authentication_test.rb]: :lollipop: (lol)

File: test/factories.rb (hunk 3 above)
Comment (qrush) [inline test/factories.rb]: It's just for a test factory and example.org/com/net/etc should be fine.

File: config/routes.rb (hunk 2 above)
Comment (qrush) [inline config/routes.rb]: It's an internal URL and the most important routes are at the top.

File: config/initializers/admins.rb (hunk 1 above)
Comment (qrush) [inline config/initializers/admins.rb]: If admins aren't setup, the app shouldn't boot. It's a configuration error from the get go. Much the same as something missing from `database.yml`.

File: test/integration/authentication_test.rb (hunk 4 above)
Comment (qrush) [inline test/integration/authentication_test.rb]: :meat_on_bone:

File: Gemfile [hunk 5]
```diff
@@ -13,12 +13,14 @@ gem 'daemons'
...
 gem 'delayed_job'
 gem 'delayed_job_active_record'
+gem 'doorkeeper'
 gem 'dynamic_form'
 gem 'gchartrb', require: 'google_chart'
 gem 'gravtastic'
 gem 'high_voltage'
 gem 'highline'
 gem 'honeybadger', '~> 2.1.0'
+gem 'http_accept_language'
```
Comment (simi) [inline Gemfile]: :scissors:

File: app/controllers/api/base_controller.rb [hunk 6]
```diff
@@ -1,8 +1,14 @@
 class Api::BaseController < ApplicationController
+  before_action :doorkeeper_authorize!, if: :doorkeeper_token
+  before_action :authenticate_with_oauth, if: :doorkeeper_token
   skip_before_action :require_ssl
 
   private
 
+  def authenticate_with_oauth
+    sign_in User.find_by_id(doorkeeper_token.resource_owner_id)
```
Comment (simi) [inline app/controllers/api/base_controller.rb]: Isn't this deprecated in favour of `User.find_by(id: doorkeeper_token.resource_owner_id)`?

**Comment by dwradcliffe** [comment]
Rebased on master and now the tests are passing. Looks good so I'm going to merge.

**Comment by bf4** [comment]
:100:

**Pull request merged**